### ISSUE TITLE
Trying to implement LARS/LAMB optimizers

### ISSUE THREAD

**Issue by harpone** [pytorch/xla#2511]
## 🐛 Bug

I'm trying to get the LAMB and LARS optimizers to work on TPUs, but I'm getting huge slowdowns at the first call of the `.step()` method. Basically something like 1-5 seconds for each `p` in `group['params']`.

See here for a [gist.](https://gist.github.com/harpone/ad7adbb7562d7cf111527ce457c4a6d6).

I tweaked it a bit to get rid of deprecation warnings for `add_` and brought it more in line with the official [adam implementation](https://github.com/pytorch/pytorch/blob/dc67b47bc9d53dbeb898a4d920b0225ac73629ec/torch/optim/adam.py#L56).

The only differences to adam I can see are 1) the `state` now contains XLA tensors and 2) there's e.g. a weight norm computation inside the `step` method (`weight_norm = p.pow(2).sum().sqrt().clamp(0, 10)`).

I can't see how these would be an issue though... is there something else going on?

## Environment
Collecting environment information...
PyTorch version: 1.6.0
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Debian GNU/Linux 9.13 (stretch) (x86_64)
GCC version: (Debian 6.3.0-18+deb9u1) 6.3.0 20170516
Clang version: Could not collect
CMake version: version 3.14.0

Python version: 3.6 (64-bit runtime)
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.1
[pip3] numpy-indexed==0.3.5
[pip3] numpydoc==1.1.0
[pip3] pytorch-lightning==0.9.1rc3
[pip3] pytorch-lightning-bolts==0.2.2
[pip3] torch==1.6.0
[pip3] torch-xla==1.6
[pip3] torchvision==0.7.0
[conda] blas                      1.0                         mkl  
[conda] mkl                       2020.3                intel_279    intel
[conda] mkl-service               2.3.0            py36he904b0f_0  
[conda] mkl_fft                   1.1.0            py36h23d657b_0  
[conda] mkl_random                1.1.1            py36h0573a6f_0  
[conda] numpy                     1.19.1           py36hbc911f0_0  
[conda] numpy-base                1.19.1           py36hfa32c7d_0  
[conda] numpy-indexed             0.3.5                     <pip>
[conda] numpydoc                  1.1.0                      py_0  
[conda] pytorch-lightning         0.9.1rc3                  <pip>
[conda] pytorch-lightning-bolts   0.2.2                     <pip>
[conda] torch                     1.6.0                     <pip>
[conda] torch-xla                 1.6                       <pip>
[conda] torchvision               0.7.0                     <pip>

 - PyTorch Version (e.g., 1.0): 1.6
 - OS (e.g., Linux): Linux deep learning VM on GCP
 - How you installed PyTorch (`conda`, `pip`, source): pre-installed
 - Build command you used (if compiling from source):
 - Python version: 3.6

**Comment by JackCaoG** [comment]
Hi @harpone could you do a debug run and post the output following the example in [here](https://github.com/pytorch/xla/blob/master/contrib/colab/issue-report.ipynb)? The metric and graph should tell us where the slowness coming from.

**Comment by harpone** [comment]
OK I'm now running the debug_run. Set it up to stop after 2 iterations, but it's going very slowly... gets stuck at `Compiling IR graph hash f253e7d28a387060367d5c01096e61a9 on device TPU:0 ...` (running on 1 core only) for several seconds, probably for each parameter `p` in `group['params']`.

BTW is it possible to disable to XLA autojit/compilation completely, at least for debugging purposes? I feel like it's a big source of pytorch-xla problems, and probably the performance gain is not huge?

**Comment by harpone** [comment]
OK the debug results are here: https://drive.google.com/file/d/1fit-YhYFl0ujrGfsWANxVrqwxSem6MHe/view?usp=sharing

I completed 2 training steps, which took some hours to complete...

**Comment by JackCaoG** [comment]
XLA is a graph base compiler, I don't think there is a way to execute the graph without a compilation 😄 . While it is annoying to wait for the compile every-time during the debugging, in the production we usually only compile very few times in the first couple steps and then keep execute the cache compiles. (Execute is usually super fast).

I am looking at your metric report and graph, for some reason it log it as 6 steps. The first 4 steps seems relative OK, there are 4 compiles and 2 cache compiles.

```
[MetricsData; step=4]
Metric: CompileTime
  TotalSamples: 4
  Accumulator: 08s849ms784.283us
  ValueRate: 415ms999.549us / second
  Rate: 0.211497 / second
  Percentiles: 1%=016ms372.081us; 5%=016ms372.081us; 10%=016ms372.081us; 20%=016ms372.081us; 50%=04s588ms508.900us; 80%=04s670ms368.276us; 90%=04s670ms368.276us; 95%=04s670ms368.276us; 99%=04s670ms368.276us
Metric: ExecuteTime
  TotalSamples: 6
  Accumulator: 141ms248.232us
  ValueRate: 007ms473.520us / second
  Rate: 0.317463 / second
  Percentiles: 1%=015ms431.172us; 5%=015ms431.172us; 10%=015ms431.172us; 20%=018ms932.291us; 50%=022ms528.131us; 80%=032ms561.932us; 90%=034ms668.513us; 95%=034ms668.513us; 99%=034ms668.513us
Counter: CachedCompile
  Value: 2
```

However during from step5 I am seeing hundreds of recompile
```
[MetricsData; step=5]
Metric: CompileTime
  TotalSamples: 500
  Accumulator: 47m60s854ms962.246us
  ValueRate: 972ms223.306us / second
  Rate: 0.172389 / second
  Percentiles: 1%=012ms608.429us; 5%=012ms367.517us; 10%=013ms886.245us; 20%=016ms774.481us; 50%=06s880ms038.364us; 80%=09s248ms852.215us; 90%=11s526ms330.388us; 95%=11s747ms208.049us; 99%=11s200ms879.540us
Metric: ExecuteTime
  TotalSamples: 797
  Accumulator: 21s670ms183.026us
  ValueRate: 007ms189.302us / second
  Rate: 0.277205 / second
Counter: CachedCompile
  Value: 298
```

I am suspecting this is the part that you mentioned to compile `for each parameter p in group['params'].`

Looking at the graph for those hundreds of recompile, it seems like they are doing the same computation but the parameter is of different size of it has to recompile
```
graph_0026
ENTRY %IrToHlo.24 (p0.1: f32[], p1.2: f32[96,1,3,3]) -> (pred[]) {
  %constant.11 = s32[] constant(864), metadata={op_type="aten::all" source_file="detect_nan_tensors@training_tricks.py" source_line=61}
....
```
```
graph_0027
ENTRY %IrToHlo.24 (p0.1: f32[], p1.2: f32[96]) -> (pred[]) {
  %constant.11 = s32[] constant(96), metadata={op_type="aten::all" source_file="detect_nan_tensors@training_tricks.py" source_line=61}
....
```
Note that `p1.2` has different size, and that's the input(tensor) to the graph. As you can see the compile time dominate the whole runtime. In order to achieve better performance with pt/xla, we need to reduce the number of recompile. Is there a way to make every parameter to be the same size (through padding for example)?

**Comment by harpone** [comment]
> XLA is a graph base compiler, I don't think there is a way to execute the graph without a compilation smile .
haha OK just checking :smile: 

OK thanks for the analysis! I think I actually did see the slowdowns starting only after maybe 5 steps or so... but these steps are not full training steps but I think each step here is a single parameter in the optimizer `.step()` method. I'd understand if the input tensors here were e.g. actual image tensors, but these are the model parameters, so padding etc. is clearly out of the question. I don't think Adam does anything like that in pt/xla??

I feel like there's something really funny going on here... is there maybe some pt/xla specific version of Adam, SGD etc used somewhere? Or does pt/xla pull some tricks for known pt optimizers such as SGD and Adam, and fails with user implemented optimizers?

**Comment by harpone** [comment]
OK I think it's the weight norm and adam norm computations inside the `.step()` method in Lamb that result in the slowdowns. If just don't use the `trust_ratio` (on line 118 [here](https://gist.github.com/harpone/ad7adbb7562d7cf111527ce457c4a6d6#file-lamb-py-L118)) like this:

```
#p.add_(adam_step, alpha=-step_size * trust_ratio)
p.add_(adam_step, alpha=-step_size)
```

Then it's all OK. The trust ratio is computed as `trust_ratio = weight_norm / adam_norm` and `weight_norm = p.pow(2).sum().sqrt().clamp(0, 10)` and `adam_norm = adam_step.pow(2).sum().sqrt()`

So I guess then since `trust_ratio` is not used, XLA will optimize it away it in the compilation step, i.e. the weight norms etc are not actually even computed?

Still don't understand why the computations are so slow though...

**Comment by jysohn23** [comment]
I see that the number of `aten::_local_scalar_dense` increases drastically in steps 4 and 5:

```
[aten::_local_scalar_dense]
0       1.0
1       2.0
2       2.0
3       2.0
4       790.0
5       1678.0
```

Are there any cpu data accesses happening in the optimizer computations or anything like that?

**Comment by harpone** [comment]
Hmm I think actually the first 4 steps could be the 2 validation steps (forward only), and then 4 and 5 do the backward pass and optimizer steps... but the number 790 sounds about like the number of parameters in the model.

But I think I found the reason(s) for the slowdowns: one was apparently the `if weight_norm == 0. or adam_norm == 0.:` check. Maybe that comparison was cast to CPU or something? I removed that.

The other was due to using a torch.tensor in `p.add_(adam_step, alpha=-step_size * trust_ratio)` (the `trust_ratio`). I changed that to 
```
adam_step.mul_(trust_ratio)
p.add_(adam_step, alpha=-step_size)
```
and now it's working! (Well at least in the sense that it's as fast as Adam now). So I guess the `alpha` kwarg must always be a non-tensor/ not on XLA device (?).

Anyone need LARS and LAMB optimizers for XLA btw? I could post them somewhere, maybe just a gist or a small repo...

**Comment by harpone** [comment]
anyway, I'll close since this is solved. Thanks all for the help!

**Comment by chandrasekhard2** [comment]
@harpone - Could you please post your fixed LARS implementation?
